### DISCUSSION TITLE
Visualising basis expansion of basex inverse abel transform

### DISCUSSION THREAD

**Discussion by David1one** [PyAbel/PyAbel#393] [General]
How may I plot the basis-expanded raw data before performing an inverse abel transform with the basex method?

From my understanding of the original basex paper, the raw data is represented as a sum of gaussians (for the forward transform at least).

![image](https://github.com/user-attachments/assets/991d362a-ea7c-47cb-9e59-1c9289ccde73)

It has been previously discussed in issue #306 how one may recover the weights of this transform.

How can they be used to visualise the basis-expanded raw data representation?

**Comment by MikhailRyazanov** [comment]
Take a look at https://pyabel.readthedocs.io/en/latest/transform_methods/basex-comp.html#intensity-correction. The functions are nor really gaussians, but are very similar (away from the axis).

(I've converted this from an “issue” to a “discussion”, since it's about how to use PyAbel rather than about it not working well.)

**Reply by MikhailRyazanov** [reply]
The code for that plot is here: https://github.com/PyAbel/PyAbel/blob/master/doc/transform_methods/basex-basis.py. You can take the basis functions from there and multiply them by the coefficients. The original paper just takes the raw data samples as the coefficients. PyAbel can do the same if you use `correction=False`.

**Reply by MikhailRyazanov** [reply]
In fact, these gaussian-like functions are used for the _transformed_ data. The raw input data (if you mean the inverse transform) is implicitly represented as a sum of the _projections_ of these gaussian-like functions. I don't have their plots, but if you really need _that_, the simplest way would be to apply a numeric forward Abel transform (with appropriate `dr`) to the functions from `basex-basis.py`, then multiply them by the expansion coefficients (from the inverse transform of the raw data).

**Reply by David1one** [reply]
Hi Mikhail. Thanks for showing me the code! Could I please ask for clarification about your second comment. The basis functions in the paper are defined as a function of r, so I believe that it is a sum of the basis functions being used to represent the 3D cylindrically symmetric object. Therefore the result of the forward transform, the 2D projection data, is represented as a sum of the projection of the basis functions. 

Therefore, if one wishes to perform a forward abel transform with basex, they must represent the 3D cylindrically symmetric object first as a sum of the gaussian-like basis functions. And for the inverse abel transform, one must start by representing the 2D projection input as a sum of the projections of the gaussian-like basis functions. 

In either case, the expansion coefficients are given by the input - the kth order basis function is multiplied by the pixel value at k=u.

Is this the right way around? And is this the correct way to construct the coefficients?

**Reply by MikhailRyazanov** [reply]
> the kth order basis function is multiplied by the pixel value at k=u

Not really. The expansion coefficients are not equal to either input or output pixel values, but rather the input and output pixel values are linear functions of the expansion coefficients according to eq. (9) and (10) in the article. So the coefficients are constructed by solving the corresponding system of linear equations for the input data. The output is then created as the linear combination of the corresponding output basis functions with these coefficients. The actual algorithm, however, doesn't use the coefficients themselves but instead combines all these linear transformations into a single matrix transform from the input to the output pixel values.

(Why do you need to “plot the basis-expanded raw data”?)

**Reply by DanHickstein** [reply]
> Why do you need to “plot the basis-expanded raw data”?

Seems potentially useful for checking the quality of the fit, right?

**Reply by MikhailRyazanov** [reply]
> potentially useful for checking the quality of the fit

Only if you need subpixel resolution. (Again, why? :grin:)

With `sigma=1` and without regularization, the input pixel values are always “fitted” exactly.
Using larger `sigma` or any regularization will indeed produce a “fit” smoother that the input data, but then if only comparison with the input pixel values is needed (without a continuous curve between the data points), then the simplest solution would be to perform the opposite transform of the result, with the same `sigma`, but without regularization and with `correction` enabled.

**Reply by David1one** [reply]
Plotting the basis set expansion is planned to be a way to double-check that the abel transform was operating as expected. It is supposed to be the same as the input 2D projection data, but it's always good to check. In addition, I need to be able to explain how the code operates, so in part this is a pedagogical exercise for me.

In addition, from my understanding, the particular fit matters - whether the basex algorithm achieves a fit similar to a linear interpolation of the data, or a comb-like structure of delta-like functions that peak at the value of the raw data, each should be expected to give different results for the retrieved 3D object radial profile.

@MikhailRyazanov thank you for your explanations. I'm slowly understanding the details of the algorithm. So if I understand correctly, the steps I need to take are:

1. Perform an inverse abel transform on my 2D projection data to retrieve the 3D data as a function of r.
2. Extend the code provided in `basex-basis.py` to calculate the correct number of basis functions necessary for my particular dataset.
3. Perform a forward abel transform on each of the functions to get their projection.
4. Use Eqn 9 & 10 in the paper to relate the inverse abel transformed raw data to the expansion coefficients - though I'm not sure how they can be used to do that. Also, aren't they in different spaces?
5. Multiply each expansion coefficient by the corresponding basis function. Plot the sum of these basis functions.

Is this correct?

**Reply by MikhailRyazanov** [reply]
> It is supposed to be the same as the input 2D projection data, but it's always good to check.

As I've said, you can always perform the opposite transform of the result. Without regularization, it will return reproduce the input data exactly (within FP errors) simply because the forward- and inverse-transform matrices are inverses of each other.

> ... whether the basex algorithm achieves a fit similar to a linear interpolation of the data, or a comb-like structure of delta-like functions that peak at the value of the raw data ...

It's definitely not delta-like because the spacing between peak centers is similar to their FWHM. And not linear, but sort of smoothly oscillating, as shown in the plot that I've mentioned above. For the raw data (the projection), the basis functions are even smoother (see $\chi_k(x)$ if the article Fig. 1; even if $\rho_k(r)$ were δ-functions, the projections will only differ by a sharp drop outside the maximum, but the inner part will look almost the same).

> should be expected to give different results for the retrieved 3D object radial profile.

To begin with, all the transform methods assume that the underlying distribution is sufficiently smooth. The [Nyquist–Shannon sampling theorem](https://en.wikipedia.org/wiki/Nyquist%E2%80%93Shannon_sampling_theorem) doesn't apply here directly, but the general idea is very similar — in order to get meaningful results, the sampling must be finer than the narrowest features in the signal.

> Is this correct?

Seems so.
Eq. (10) is in the _xz_ space. You can actually ignore the _z_ dependence and the corresponding _m_ index everywhere because separate rows are in fact transformed independently (the article does transform them, but the overal effect is an identity transform). Eq. (10) without _z_ and _m_ will be $P(x_i) = \sum_{k=0}^{K_x-1} C_k \chi_k(x_i).$ To get the continuous form, just use your $x$ variable instead of the sampling points $x_i$.

P.S. If you're not tied to BASEX specifically, I'd suggest to look at the Daun method, which is based on the same ideas but uses [basis functions](https://pyabel.readthedocs.io/en/latest/transform_methods/daun.html#basis-sets) with more reasonable properties and much easier to calculate.

**Reply by David1one** [reply]
Hi Mikhail,

Thank you for your explanation! I will have to take some time to digest it, and I'll follow up when appropriate.

Warm regards,
David
